skorch/net.py

```diff
@@ -567,56 +540,214 @@ def _apply_virtual_params(self, virtual_kwargs):
+                    if optimizer_name == 'optimizer':
+                        yield from self.module_.named_parameters()
+                    else:
+                        yield from self.module2_.named_parameters()
+
+                def initialize_optimizer(self, *args, **kwargs):
+                    super().initialize_optimizer(*args, **kwargs)
+
+                    # add an additional optimizer called 'optimizer2_' that is
+                    # only responsible for training 'module2_'
+                    named_params = self.get_learnable_params('optimizer2')
+                    args, kwargs = self.get_params_for_optimizer(
+                        'optimizer2', named_params)
+                    self.optimizer2_ = torch.optim.SGD(*args, **kwargs)
+                    return self
+
+        Parameters
+        ----------
+        optimizer_name : str (default='optimizer')
+          The name of the optimizer that will be responsible for updateing these
+          parameters. By default, this argument is not used in the method body
+          but it can be useful if you choose to override this method.
+
+        Yields
+        ------
+        named_parameters
+
+          An iterator over all module parameters, yielding both the name of the
+          parameter as well as the parameter itself. Use this, for instance, to
+          pass the named parameters to :meth:`.get_params_for_optimizer`.
 
         """
-        self.initialize_virtual_params()
-        self.initialize_callbacks()
-        self.initialize_criterion()
-        self.initialize_module()
-        self.initialize_optimizer()
-        self.initialize_history()
+        for name in self.modules_ + self.criteria_:
+            module = getattr(self, name + '_')
+            named_parameters = getattr(module, 'named_parameters', None)
+            if named_parameters:
+                yield from named_parameters()
+
+    def _initialize_optimizer(self, reason=None):
+        with self._current_init_context('optimizer'):
+            named_parameters = self.get_learnable_params('optimizer')
+            _, kwargs = self.get_params_for_optimizer(
+                'optimizer', named_parameters)
+
+            if self.initialized_ and self.verbose:
+                if reason:
+                    # re-initialization was triggered indirectly
+                    msg = reason
+                else:
+                    # re-initialization was triggered directly
+                    msg = self._format_reinit_msg(
+                        "optimizer", kwargs, triggered_directly=False)
+                print(msg)
+
+            self.initialize_optimizer()
+
+            self._register_virtual_param(
+                ['optimizer__param_groups__*__*', 'optimizer__*', 'lr'],
+                optimizer_setter,
+            )
+            return self
+
+    def _initialize_history(self):
+        with self._current_init_context('callbacks'):
```

Comment:
I do not think we actually explicitly use of the `'callbacks'` or `virtual_params` contexts. Should we still have them?

Reply:
Good question. My reasoning was to stay consistent, but I don't see a use case at the moment. I could leave the context managers there but add a comment that they're not used at the moment.